Bug 1922986 - Create script that exports BMO data as JSON suitable for import into a BigQuery instance in GCP

dklawren opened the pull request:
Sorry this so late. Lots of moving pieces had to be coordinated with webservices-infra, etc. just so I could test and then small changes one after another to get to this point. 

- Added config options for BMO ETL settings such as base url, project id, dataset id, etc.
- A export script that will run nightly at 11pm and export daily snapshot of specific bug data to BigQuery
- A test BQ emulator and test script that can be used to do basic testing of exporting bug data to the emulator locally. There is a data.yml file that creates the same schema as what will be in production before exporting.
- Small change to circleci and github ci to run the tests in the extensions directory.

Let me know if you have any questions as it is a lot.

cgsheeh commented on extensions/BMO/template/en/default/hook/admin/params/editparams-current_panel.html.tmpl:
```diff
@@ -18,5 +18,12 @@ ELSIF panel.name == "bugfields";
+ELSIF panel.name == "reports";
+  panel.param_descs.bmo_etl_enabled = 'Enable export to BMO ETL.';
+  panel.param_descs.bmo_etl_base_url = 'The base URL for sending BMO ETL data.';
+  panel.param_descs.bmo_etl_service_account = 'THe Google service account for accessing the BMO ETL API.';
```
```suggestion
  panel.param_descs.bmo_etl_service_account = 'The Google service account for accessing the BMO ETL API.';
```

cgsheeh commented on extensions/BMO/bin/export_bmo_etl.pl:
```diff
@@ -0,0 +1,750 @@
+  }
+
+# Google Kubernetes allows for the use of Workload Identity. This allows
+# us to link two serice accounts together and give special access for applications
```
```suggestion
# us to link two service accounts together and give special access for applications
```

cgsheeh commented on extensions/BMO/bin/export_bmo_etl.pl:
```diff
@@ -0,0 +1,750 @@
+  return decode_json($data);
+}
+
+sub store_cache {
```
Should we clear the cache at the end of the script?

cgsheeh reviewed (changes requested):
Looks good to me, just a few comments/questions.

It's a lot less scary in terms of PR size once you realize there is quite a lot of repetition. You could abstract a lot of that away, but if you're pressed for time/deadlines it's okay to leave as-is.

dklawren commented on docker-compose.yml:
```diff
@@ -136,17 +138,43 @@ services:
       - bmo-gcs-sitemapindex:/data/sitemapindex
       - bmo-gcs-mining:/data/mining
 
+  bigquery:
+    platform: linux/x86_64
+    image: ghcr.io/goccy/bigquery-emulator:latest
+    ports:
+      - 9050:9050
+    volumes:
+      - bmo-bigquery-data:/work
+      - ./docker/bigquery/data.yaml:/work/data.yaml
+    working_dir: /work
+    command: |
+      --project=test --data-from-yaml=/work/data.yaml --log-level=debug
+
   externalapi.test:
     platform: linux/x86_64
     build: *bmo_build
     entrypoint: perl /app/external_test_api.pl daemon -l http://*:8001
     ports:
       - 8001:8001
 
+  bq:
+    platform: linux/x86_64
+    build:
+      context: ./docker/bigquery
+      dockerfile: Dockerfile
+    ports:
+      - 9050:9050
+    volumes:
+      - bmo-bq-data:/work
+    working_dir: /work
+    command: |
+      --project=test --data-from-yaml=/data.yaml --log-level=debug
+
```
bq is the correct one and bigquery is there by mistake and will be removed.

dklawren commented on extensions/BMO/bin/export_bmo_etl.pl:
```diff
@@ -0,0 +1,750 @@
+        classification  => $obj->classification,
+        comment_count   => $obj->comment_count,
+        vote_count      => $obj->votes,
+        group           => (join ',', map { $_->name } @{$obj->groups_in}),
```
Didn't see that before. I was originally going by the schema here https://docs.google.com/spreadsheets/d/1LDrYkGxZGnktI3-Ei3fpPK2TAXjxTj7fBUjHaJ8tpBE/edit?gid=0#gid=0 which called for the group names to just be a single string. I will need to file a couple of PRs to get the schema changed but better to do so now before we have actual data going in.

dklawren replied:
https://github.com/mozilla-it/webservices-infra/pull/3627

dklawren commented:
I need to make some other non-related changes anyway as I realized i left out two additional requirements. 1) I need to skip sending bug data related to the 'Legal' product and 2) I need to filter the summaries, etc. for bugs that are private.

https://docs.google.com/document/d/1D-0wLQCrYXw17k_qSZiaH8djaUV-iS3qUencaj0FTtc/edit?tab=t.0#heading=h.l46cjmt1mwn6

dklawren commented:
I am going to make one last schema change to the BQ data so I will need to create the needed PRs in webservices-infra for this :( Once that is done, I will be able to submitted the updated code here for last review.

Context: The bugs.group field needs to be a string as I originally had it as 1) most bugs do not have more than one group anyway and 2) if there is more than one group, we just want to put the group name with the least amount of users. I missed that little detail from the schema outlined here: https://docs.google.com/spreadsheets/d/1LDrYkGxZGnktI3-Ei3fpPK2TAXjxTj7fBUjHaJ8tpBE/edit

Sorry for the extra delay.

dklawren commented:
Ready for hopefully final review. Optimized as much as I could. Refactored to create functions where possible. Switched to using LIMIT/OFFSET to allow for breaking down the tables that do not have a primary key into chunks instead of loading the whole table into memory (i.e. dependencies, etc.). Let me know if anything needs to be elaborated on.
